### PR TITLE
feat(ops): PR-F5 — preflight hardening (DAG conf validation + doctor --memory)

### PR DESCRIPTION
**Pull request by Kopanov**
## Summary

Two complementary operator-visibility additions surfaced by Bravo's 2026-04-19 / 2026-04-20 overnight-baseline investigation. Both are about closing **silent-failure gaps** that bit production in the past week.

### (1) Baseline DAG conf typo / unknown-key validation

**The 2026-04-19 incident:** an operator triggered a fresh baseline via the Airflow UI with `{"days": 730}` instead of `{"baseline_days": 730}` AND missing `fresh_baseline: true` entirely. Both consumers silently fell through to defaults — additive mode (no wipe), default 730-day window. The misconfiguration was only caught hours later in postmortem.

**The fix:** new `_validate_baseline_dag_conf` helper called from EVERY conf-consuming function (3 sites: `get_baseline_config`, `_baseline_clean`, `_baseline_start_summary`). For each unrecognized key it emits a WARNING log line with grep marker `[BASELINE_CONF]`; common typos get a "did you mean?" suggestion (e.g. `days` → `baseline_days`, `fresh` → `fresh_baseline`). Known-good keys documented in `docs/AIRFLOW_DAGS.md` as a single canonical reference.

WARNING (not ERROR) is deliberate: refusing to run on an unknown key would break operators who add forward-compat keys ahead of code that consumes them. Operators / scripted CI can grep the marker `[BASELINE_CONF]` to fail loudly; humans see the warning in the log and reorient.

### (2) `edgeguard doctor --memory` diagnostic

**The visibility gap:** PR-F4 (#60) addressed the symptom of MISP HTTP 500s by sequencing tier-1 collectors. But the underlying memory-tuning recommendations (Neo4j tx-memory bumped 4G→8G in 2026-04, MISP PHP `memory_limit` 1G+) were buried in source comments — operators couldn't see whether their installation was misconfigured before triggering a multi-hour baseline.

**The fix:** new `--memory` flag on `edgeguard doctor`. Probes the live Neo4j container for heap / page-cache / tx-memory env vars, parses host RAM from `/proc/meminfo` or `sysctl`, compares each against recommended values, renders a verdict table (✓ ok / ⚠ low / ✗ too low / ? unknown).

Recommendations come from real incidents, not generic advice — see `docs/MEMORY_TUNING.md` for the full rationale + incident links.

Sample output:

```
Memory & sizing check

  Setting                              Current     Min     Rec  Verdict
  --------------------------------    ----------  ------  ------  -------
  Neo4j heap                                8.0G    4G      8G    ✓ ok
  Neo4j page cache                          2.0G    2G      4G    ⚠ low
                                    → Caches Neo4j store files; 4G recommended for 350K-node graph
                                    → set NEO4J_server_memory_pagecache_size in .env / docker-compose.yml
  Neo4j tx memory                           8.0G    4G      8G    ✓ ok

  Host RAM                                 16.0G    8G     16G    ✓ ok

ℹ MISP PHP settings (memory_limit, max_execution_time) are NOT auto-probed yet.
ℹ Manual check: ``docker compose exec misp php -r 'echo ini_get(\"memory_limit\");'``
ℹ Recommended values + rationale: see docs/MEMORY_TUNING.md
```

## Why these two together

Both are **operator preflight visibility** — surfacing things that today are invisible until a multi-hour baseline run goes wrong. Same testing pattern (pure helpers + source-pin + doc traceability), same quality bar, same shipping cost.

## What this PR does NOT do

The deeper architectural issues from Bravo's investigation are **deliberately out of scope** for PR-F5 — they need design discussion, not a quick fix:

- **Orphan collector process cleanup** — see [Issue #65](../../issues/65). The 2026-04-19 incident: a failed DAG run kept its `collect_nvd` Python subprocess alive for 12+ hours, eventually pushing 78,313 attributes to MISP at 01:08 UTC after the manual run's `baseline_clean` had already wiped MISP. This is a real design gap (Airflow doesn't kill task subprocesses on DAG failure) with multiple safe approaches; needs its own design discussion + PR.
- **Cross-event CVE deduplication** — [Issue #61](../../issues/61) already covers this (event partitioning by date range). Bravo's measured evidence (72,479 CVEs duplicated between events 19 + 20, ~92.6% of event 19's CVEs) has been added to that issue as concrete validation.

## Files

| File | Change |
|---|---|
| `dags/edgeguard_pipeline.py` | `_validate_baseline_dag_conf` helper + 3 call sites + KNOWN_KEYS frozenset + TYPO_MAP |
| `src/edgeguard.py` | `--memory` flag, `_doctor_memory_check`, recommendations table, pure helpers, probe helpers |
| `docs/MEMORY_TUNING.md` | **new** — single reference for "what should this be set to and why" |
| `docs/AIRFLOW_DAGS.md` | New "Baseline DAG conf — accepted keys (PR-F5)" subsection |
| `tests/test_pr_f5_dag_conf_validation.py` | **new** — 11 tests (source-pin + behavior + doc traceability) |
| `tests/test_pr_f5_doctor_memory.py` | **new** — 30 tests (argparse, parser matrix, verdict matrix, recommendations contract, probe degradation, doc traceability) |

## Test plan

- [x] DAG validator wired at all 3 conf-consumption sites (source-pin)
- [x] Known keys never warn
- [x] Typo `days` → "did you mean `baseline_days`?" suggestion fires
- [x] Unknown unmapped keys get generic warning with known-keys list
- [x] Empty / None / non-dict conf is a safe no-op (no crash, no spurious warnings)
- [x] Case + whitespace tolerant typo lookup (`"  Days  "` still suggests)
- [x] `--memory` flag registered on doctor subparser
- [x] `_parse_memory_value_to_gb` handles `4G`, `512M`, `1024k`, `2T`, bare bytes, whitespace, unparseable
- [x] `_verdict_for_memory` matrix: ok / warn / fail / unknown across thresholds
- [x] Neo4j tx-memory recommendation pinned at 8G (the 2026-04-18 incident value)
- [x] Probes degrade gracefully when docker / `/proc/meminfo` / `sysctl` unavailable
- [x] `MEMORY_TUNING.md` references `MemoryLimitExceededException` symptom + PR-F4 incident
- [x] `ruff format --check` ✓
- [x] `ruff check` ✓
- [x] `mypy src/ scripts/ --ignore-missing-imports` ✓
- [x] `pytest tests/` — **865 passed, 3 skipped, 0 failures** (was 833 — +32 new tests)

## Related

- PR #60 (PR-F4) — sequenced tier-1 collectors (MISP HTTP 500 fix; surfaced the memory-tuning visibility gap)
- Issue #57 — Airflow-aware baseline lock primitive (companion design issue)
- Issue #61 — MISP event partitioning (Bravo's 72K-duplication evidence added)
- Issue #65 — orphan collector process cleanup (the deepest architectural gap from Bravo's investigation)
- Bravo's investigation — Slack threads 2026-04-19 22:47 UTC onward and 2026-04-20 11:52 AM (Brussels) onward

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Low Risk**
> Changes are additive and primarily logging/diagnostics plus documentation/tests; the main risk is noisy warnings or subprocess probing edge cases, not data-path behavior.
> 
> **Overview**
> Adds operator-facing preflight visibility in two places.
> 
> In `edgeguard_baseline`, introduces `_validate_baseline_dag_conf` to **warn on any unrecognized `dag_run.conf` keys** (with `[BASELINE_CONF]` marker and “did you mean …” suggestions for common typos) and wires it into all current conf-consumption sites (`get_baseline_config`, `_baseline_start_summary`, `_baseline_clean`) to prevent silent fallback during destructive/additive baseline runs.
> 
> Extends the CLI with `edgeguard doctor --memory`, which probes Neo4j container memory env vars and host RAM, parses values, compares them against an incident-driven recommendations table, and prints a verdict report; adds new ops docs (`docs/MEMORY_TUNING.md`), updates `docs/AIRFLOW_DAGS.md` to document accepted baseline conf keys, and adds regression tests covering wiring, behavior, and doc traceability.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 37323aefec04df73ffbf36b3ea22a2d1549e4a64. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/edgeguard.py [hunk 1]
```diff
@@ -550,6 +560,242 @@ def cmd_doctor(args):
...
+    # Linux first
+    try:
+        with open("/proc/meminfo") as fh:
+            for line in fh:
+                if line.startswith("MemTotal:"):
+                    parts = line.split()
+                    if len(parts) >= 2:
+                        kb = float(parts[1])
+                        return round(kb / (1024**2), 1)
+    except (FileNotFoundError, OSError):
```
[the remark this answers is not in the document]
Comment (Kopanov) [reply src/edgeguard.py]: Resolved in commit `95401f7`. Real inconsistency — the Linux `/proc/meminfo` path had `(FileNotFoundError, OSError)` while the macOS `sysctl` path correctly included `ValueError`. The asymmetry violated the function's pinned "MUST return None on any failure — never raise" contract.

Fix: add `ValueError` to the Linux except tuple. Brings parity with the macOS path. New regression test `test_host_ram_probe_returns_none_on_malformed_meminfo` simulates malformed meminfo content (`MemTotal: garbage kB`) AND a failing sysctl fallback to verify the full graceful-degradation chain returns `None` without raising.

44 PR-F5 tests pass (was 41, +3 new from this and the verdict_glyph fix).

File: src/edgeguard.py [hunk 2]
```diff
@@ -550,6 +560,242 @@ def cmd_doctor(args):
...
+        # exec, where the container's own env vars are visible).
+        raw = neo4j_env.get(env_var) or os.environ.get(env_var)
+        current_gb = _parse_memory_value_to_gb(raw)
+        verdict = _verdict_for_memory(current_gb, rec["min_gb"], rec["rec_gb"])
+        verdict_glyph = {
+            "ok": f"{Colors.GREEN}✓ ok{Colors.END}",
+            "warn": f"{Colors.YELLOW}⚠ low{Colors.END}",
+            "fail": f"{Colors.RED}✗ too low{Colors.END}",
+            "unknown": f"{Colors.YELLOW}? unknown{Colors.END}",
+        }[verdict]
```
[the remark this answers is not in the document]
Comment (Kopanov) [reply src/edgeguard.py]: Resolved in commit `95401f7`. Real DRY violation — the identical 4-entry verdict→glyph dict was inlined twice in `_doctor_memory_check` (once for the Neo4j-settings loop, once for the host-RAM section). Changing the wording or colors in one but not the other would silently produce inconsistent output.

Fix: extract to a module-level pure helper `_verdict_glyph(verdict: str) -> str` that both call sites use. Defensive fallthrough returns the raw input string for unexpected verdicts (no crash on programmer error). 2 new tests cover the four known verdicts + the defensive fallthrough.

44 PR-F5 tests pass.

File: tests/test_pr_f5_dag_conf_validation.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        with open("docs/AIRFLOW_DAGS.md") as fh:
+            content = fh.read()
+        # The new "Baseline DAG conf — accepted keys (PR-F5)" subsection
+        # must exist with all known keys + the [BASELINE_CONF] grep marker.
+        assert "Baseline DAG conf — accepted keys" in content
+        assert "[BASELINE_CONF]" in content, (
+            "docs must mention the [BASELINE_CONF] log marker so operators know what to grep for"
+        )
+        for key in ("fresh_baseline", "baseline_days", "collection_limit", "clear_checkpoints"):
+            assert f"`{key}`" in content, f"docs must document the {key!r} conf key"
```
[the remark this answers is not in the document]
Comment (Kopanov) [reply tests/test_pr_f5_dag_conf_validation.py]: Resolved in commit `37323ae`. Real drift between the test's hardcoded 4-key tuple and the 5-key SSoT.

Fix: iterate `_KNOWN_BASELINE_CONF_KEYS` directly via the existing `_import_validator()` helper (which already loads the frozenset). Now the test automatically covers any future additions/removals to the SSoT — the contract "docs must list all accepted keys" can never silently break again.

`baseline_collection_limit` is already in `docs/AIRFLOW_DAGS.md` line 206 (legacy alias for `collection_limit`, kept for back-compat); the gap was only in test coverage, not the docs themselves.

12 PR-F5 conf-validation tests pass.
